psh/test-ls.py

```diff
@@ -0,0 +1,348 @@
+                 + r'(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)( )'
+                 + r'[1-31]( *)'
+                 + r'(\d){2}(:)(\d){2}( )'
+                 + r'(\x1b\[34m)?'
+                 + f'{file}')
+
+    p.expect(EOL + PROMPT)
+
+
+def assert_ls_r(p, files):
+    cmd = 'ls -r testDir'
+    exp_list = ''
+    hidd_files_nr = 1
+    for file in sorted(files, reverse=True):
+        if file == 'dir':
+            exp_list += esc_codes['set_blue']
+        if file[0] == '.':
+            continue
+        # last position
+        if file == sorted(files, reverse=True)[len(files)-1-hidd_files_nr]:
+            exp_list += f'{file}' + esc_codes['reset_attributes']
+        else:
+            exp_list += f'{file}  ' + esc_codes['reset_attributes']
+
+    p.sendline(cmd)
+    p.expect(cmd + EOL + exp_list + EOL + PROMPT)
+
+
+def assert_ls_S(p, files):
+    cmd = 'ls -S testDir'
+    # dir size > empty files size
+    exp_list = esc_codes['set_blue'] + 'dir  ' + esc_codes['reset_attributes']
+
+    for file in sorted(files):
+        if file == 'dir':
+            continue
+        if file[0] == '.':
+            continue
+        if file == sorted(files)[len(files)-1]:
+            exp_list += f'{file}' + esc_codes['reset_attributes']
+        else:
+            exp_list += f'{file}  ' + esc_codes['reset_attributes']
+
+    p.sendline(cmd)
+    p.expect(cmd + EOL + exp_list + EOL + PROMPT)
+
+# newly created files have all 'Jan 1 00:00' creation time
+# so can't check if it works properly in automatic test
+# def assert_ls_t(p, files):
+#     cmd = 'ls -f testDir'
+#     exp_list = ''
+#     for file in files:
+#         if file == 'dir':
+#             exp_list += esc_codes['set_blue']
+#         if file[0] == '.':
+#             continue
+#         # last position
+#         if file == files[len(files)-1]:
+#             exp_list += f'{file}' + esc_codes['reset_attributes']
+#         else:
+#             exp_list += f'{file}  ' + esc_codes['reset_attributes']
+
+#     p.sendline(cmd)
+#     p.expect(cmd + EOL + exp_list + EOL + PROMPT)
+
+
+def assert_ls_devcolor(p):
+    cmd = 'ls dev'
+    # the following regex will match list of various devices
+    exp_list = f'(({esc_codes["set_yellowonblack"]}' \
+               + r'(\S+)[ \t]*' \
+               + f'{esc_codes["reset_attributes"]}' \
+               + r')+' \
+               + EOL \
+               + r')+'
+
+    p.sendline(cmd)
+    p.expect(exp_list + PROMPT)
+
+
+def assert_ls_execolor(p):
+    binaries_dir = binaries_dir_is(p)
+    cmd = f'ls {binaries_dir}'
+
+    p.sendline(cmd)
+    # assume that ls binary is in directory with binaries
+    p.expect(esc_codes['set_green'] + 'ls')
+
+
+def assert_ls_pshcmds(p):
+    binaries_dir = binaries_dir_is(p)
+    cmd = f'ls {binaries_dir}'
+    exp_list = ''
+
+    for pshcmd in PSH_COMMANDS:
+        exp_list = exp_list + esc_codes['set_green'] + pshcmd + esc_codes["reset_attributes"] + EOL + '?'
```

Comment:
this is invalid as it assumes there are ONLY psh commands in the result. When You would build the full project there will be for example `busybox` binary between `bind` and `cat`.

Also - I would add param to disable coloring in psh `ls` (the same way it's done in buxybox ls) to make the parsing in tests easier.

Reply:
That's right, now it would works well and check if all commands from PSH_COMMANDS list are visible (even if between them are other files)

So do You think it's better to disable coloring when testing and don't test this functionality? Tests, which parse colored output are already written, I would leave leave it as it is. However, I agree param like that in ls would be a good idea.